lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -1190,6 +1194,17 @@ void FIRRTLModuleLowering::lowerModuleBody(
       continue;
     }
 
+    // Ports of opaque type *have* to be materialized directly by the code
```

Comment:
Do we need opaque types to wrap unknown types, or is it enough to say !isa<FIRRTLType> and do this logic?

Reply:
This was also my first thought: just make sure that non-FIRRTL types are passed through as-is. The problem I came across when I tried to implement that was the FIRRTL dialect and associated passes (all the width/reset inference, IMConstProp, type lowering, etc.) making ample use of `type.cast<FIRRTLType>()` and friends.

I'm not quite sure what to do here. On one hand, supporting foreign types directly would be a lot nicer and more ergonomic (no `!firrtl.opaque` wrapping). But on the other hand, this requires undoing the assumption baked into a lot of passes that types are FIRRTL types.

On a side-note: I think the only ops that need foreign type support are `WireOp`, `FModuleOp`, `InstanceOp`, `StrictConnectOp`. Anything besides passing foreign values through the module hierarchy doesn't strike me as an immediate use case.

I'm happy to put in the work, just went for the easier approach first :stuck_out_tongue:. What's your preference @darthscsi?